Erlay refactored

naumenkogs opened the pull request:
I'm using this PR to get an initial review on the new structure of the Erlay PR.

*please make it high level and keep low-level comments for the main PR later*

Note that this is missing some comments, but it should be sufficient for now to evaluate my modularity approach.

My own questions here are:
~~1. Potenitally too much locks (see m_states_mutex in ReconHandler)? Or is that fine?~~
2. ReconPhase state machine updated from ReconHandler, and not from ReconState while doing other manipulations?
~~3. See my own [review comment](https://github.com/naumenkogs/bitcoin/pull/2#discussion_r585564153)~~

Also, I remember @sipa suggested we have TxAnnouncement separate module. I'm pretty sure it's orthogonal to this PR. This would wouldn't change much if we also expand TxReconciliationHandler to be TxAnnouncementHandler or something like that.

naumenkogs committed:
Add minisketch dependency

naumenkogs committed:
tests: support (inbound) connecting to mininode

naumenkogs committed:
Add a function to announce transactions after reconciliation

Have a separate function to announce transactions to a peer (via INVs)
considering INV message limitations.

naumenkogs committed:
Add helper to compute reconciliation tx short id

Short ids are used to compute reconciliation sketches.

naumenkogs committed:
Add helper to compute sketches for tx reconciliation

naumenkogs committed:
Add a finalize incoming reconciliation function

This currently unused function is supposed to be used once
a reconciliation round is done. It cleans the state corresponding
to the passed reconciliation.

naumenkogs committed:
Add a function to identify local/remote missing txs

When the sketches from both sides are combined successfully,
the diff is produced. Then this diff can (together with the local txs)
be used to identified which transactions are missing locally and remotely.

naumenkogs committed:
Add a function to get wtxids by shortids

At the end of a reconciliation round, a peer may ask us
for transactions by their short id. Add a function for a local
lookup short_id->wtxid.

naumenkogs committed:
Request extension if decoding failed

If after decoding a reconciliation sketch it turned out
to be insufficient to find set difference, request extension.
Store the initial sketches so that we are able to process
extension sketch.

naumenkogs committed:
Handle reconciliation finalization message

Once a peer tells us reconciliation is done, we should behave as follows:
- if it was successful, just respond them with the transactions they asked
  by short ID.
- if it was a full failure, respond with all local transactions from the reconciliation
  set snapshot
- if it was a partial failure (only low or high part was failed after a bisection),
  respond with all transactions which were asked for by short id,
  and announce local txs which belong to the failed chunk.

naumenkogs commented on src/net_processing.cpp:
I'm a little concerned about this thing here, but I'm really unsure how to make it better. Do you think it's fine? That's why we have const modifier, right? :)

jnewbery commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+     */
+    Minisketch ComputeSketch(uint16_t capacity, bool use_snapshot=false)
+    {
+        Minisketch sketch;
```
You've got a lot of implementation details in this header file. That's bad for compilation times (this'll get compiled for every file that includes txreconciliation.h directly or indirectly). I suggest you move these function implementations into the cpp file.

jnewbery commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+    void HandleIncomingExtensionRequest(const NodeId peer_id);
+
+
+    Optional<ReconciliationState> GetPeerState(const NodeId peer_id) const
```
This is piercing the encapsulation. The client should never handle a `ReconciliationState` object. There should be a class interface function in `TxReconciliationTracker` for the client to get the data that it needs.

jnewbery commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+
+    TxReconciliationTracker() {};
+
+    /**
+     * TODO: document
+     */
```
These interface function documents are the most important part of this refactor! I suggest you fill them to fully describe what interface this module is presenting to the client (net_processing). Once that's done, reviewers will be able to assess whether you're proposing a sensible interface.

jnewbery commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+ * sequence, nodes are supposed to exchange transactions, so that both of them
+ * have all relevant transactions. For more protocol details, refer to BIP-0330.
+ */
+class ReconciliationState {
```
What do you think about using a pimpl model (like txrequest)? That way, the data members of `TxReconciliationTracker` could be placed in the implementation file, including `m_states`. `ReconciliationState` could be moved entirely into the cpp file.

jnewbery commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+     * Also, salts should be different per-connection to prevent halting
+     * of relay of particular transactions due to collisions in short IDs.
+     */
+    Mutex m_local_salts_mutex;
```
Why multiple mutexes here? I think one mutex for the entire class would suffice.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2607,6 +2685,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            // Reconciliation is supported only when wtxid relay is supported for only
+            // those connections which (at least might) support transaction relay.
+            if (pfrom.MightSupportTransactionRelay()) {
+                auto recon_suggestion = m_reconciliation.SuggestReconciling(pfrom.GetId(), pfrom.IsInboundConn());
```
A structured binding would be nice here!

jnewbery commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+     * details of our local state, which should be communicated to the peer so that they better
+     * know what we need.
+     */
+    Optional<std::pair<uint16_t, uint16_t>> MaybeRequestReconciliation(const NodeId peer_id);
```
Use `std::optional` everywhere.

jnewbery commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+    /**
+     * TODO document
+     */
+    Optional<std::vector<uint8_t>> MaybeRespondToReconciliationRequest(const NodeId peer_id);
```
What's the difference for the caller between an empty vector and a nullopt here?

jnewbery reviewed (commented):
Looks to be a great improvement. The encapsulation isn't quite complete yet, but it's a good start.

ariard commented on src/net_processing.cpp:
```diff
@@ -4021,6 +4115,72 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+    // If received from outgoing, add the peer to the reconciliation queue.
+    // Feature negotiation of tx reconciliation should happen between VERSION and
+    // VERACK, to avoid relay problems from switching after a connection is up.
+    if (msg_type == NetMsgType::SENDRECON) {
```
What do you think about moving this negotiation message following other ones like `SENDHEADERS`, `SENDADDRV2`, ... ? Also if you want this message to follow the same semantic, you should disconnect peers sending `SENDRECON` after a `VERACK` has been received (`fSuccessfullyConnected`)

ariard commented on src/net_processing.cpp:
```diff
@@ -4659,24 +4819,26 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                 bool fSendTrickle = pto->HasPermission(PF_NOBAN);
                 if (pto->m_tx_relay->nNextInvSend < current_time) {
                     fSendTrickle = true;
-                    if (pto->IsInboundConn()) {
```
I think the diff is smaller if you just check !recon_state on this if statement.

ariard commented on src/net_processing.cpp:
```diff
@@ -4787,12 +4959,56 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
+                    if (recon_state && txs_to_reconcile.size() != 0) {
+                        std::vector<uint256> txs_to_flood;
+                        int32_t recon_set_overflow = recon_state->GetLocalSetSize(false) + txs_to_reconcile.size() - MAX_PEER_TX_ANNOUNCEMENTS;
+                        if (recon_set_overflow > 0) {
```
Code style, I think this expression might be clearer that way :

```
size_t recon_set_size = recon_state->GetLocalSetSize(false) + txs_to_reconcile.size()
if (recon_set_size > MAX_PEER_TX_ANNOUNCEMENTS) {}
```

naumenkogs replied:
The problem is that we will need recon_set_overflow twice later in logic. Are you suggesting re-declaring it inside the if?

ariard commented on src/net_processing.cpp:
```diff
@@ -4751,9 +4914,18 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
+                        } else {
+                            // Storing to populate the reconciliation set.
+                            txs_to_reconcile.push_back(txinfo.tx->GetWitnessHash());
+                        }
```
I think you can refactor this code area a bit:
* Substitute `vInv` by a`txs_to_announce` (`std::vector<uint256>`)
* when `vInvTx` is empty or `INVENTORY_BROADCAST_MAX` is reached exit transactions to announce selection
* if peer requires reconciliation, check if its reconciliation set isn't at capacity and drain `txs_to_announce` accordingly
* flood the remainder of `txs_to_announce`, splitting messages accordingly to `MAX_INV_SZ`

That way you conserve announcement pre-checks (feefilter, still-present-in-mempool, ...) in common across tx-announcement strategies while keeping the code path straighter.

ariard commented on src/txreconciliation.cpp:
```diff
@@ -0,0 +1,336 @@
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <txreconciliation.h>
+
```
Should you introduce a namespace ? I think it would let you drop some `static` qualifiers around.

ariard commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+        return m_requestor;
+    }
+
+    bool IsResponder() const
```
I think those 4 functions (with `GetLocalSetSize`) should be public. The remaining ones aren't used in`net_processing` and can be accessed by `TxReconciliationTracker` or its implementation by making it a friend of `ReconciliationState` ?

ariard commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+    std::chrono::microseconds GetNextRespond() const
+    {
+        return m_next_recon_respond;
+    }
```
I think you can save a lot of those helpers by definitely making `TxReconciliationTracker` a friend of `ReconciliationState` ?

ariard commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,585 @@
+        LOCK(m_local_salts_mutex);
+        m_local_salts.erase(peer_id);
+        LOCK(m_states_mutex);
+        m_states.erase(peer_id);
```
You can drain those 3 function implementations to the cpp file.

ariard reviewed (commented):
Overall I belive this is moving in the good direction! Still need a bit more work around encapsulation as underscored by John.

1. Yes I'm thinking that `m_queue_mutex` is redundant to `m_states_mutex`. Everytime your access the queue, it sounds you're also accessing per-peer reconciliation states
2. Can you precise your question further ? Grepping ReconPhase or ReconHandler doesn't return anything
3. What's your exact concern there ? You're not using `txs_to_reconcile` further so I don't see what `const` is preventing of ?

naumenkogs committed:
Squashed 'src/minisketch/' content from commit 80481500c

git-subtree-dir: src/minisketch
git-subtree-split: 80481500cdb3328adab751db30a3fa7e76fe1147

naumenkogs committed:
Merge commit '3a7c4423017bf718cbcd8415fe1c0d7fec38c142' as 'src/minisketch'

naumenkogs committed:
Add helper to recompute reconciliation Q

When a reconciliation is over, we sometimes want to
update the q coefficient to make better set difference
estimation.

naumenkogs committed:
Move RelayTransaction to PeerManager

We will need this later to choose transaction announcement
strategy based on where it came from.

naumenkogs committed:
Add certain transaction to reconciliation sets

Add transactions which are eligible for reconciliation to the
reconciliation sets of the peers, otherwise flood.

naumenkogs committed:
Refactor ProcessOrphanTx

This will be later useful for choosing announcement strategy

naumenkogs committed:
Functional tests of transaction reconciliation

naumenkogs committed:
Add helper to see if reconcil peer is flood_to

naumenkogs committed:
Reduce tx broadcast interval

Since reconciliation is naturally slower than flooding, and also
since we limit the flooding aspect, we should reduce the intervals
between broadcasts (flooding out or adding to reconciliation sets).
Otherwise, transaction relay will be too slow
(and probably less inefficient).

Note that for privacy reasons the ratio between inbound and outbound
delays matter much more than the actual delays. That ratio is preserved
here, so it is not a privacy degradation.

naumenkogs committed:
Initiate reconciliation

ariard commented on src/net_processing.cpp:
```diff
@@ -1658,17 +1700,35 @@ void PeerManagerImpl::SendPings()
 {
-    connman.ForEachNode([&txid, &wtxid](CNode* pnode) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+    bool flood;
+    if (pfrom) {
```
I think you can pass the `NodeId` directly here. You don't use `pfrom` beyond indexing in reconciliation struct ?

ariard commented on src/txreconciliation.cpp:
```diff
@@ -0,0 +1,893 @@
+
+/** Current protocol version */
+constexpr uint32_t RECON_VERSION = 1;
+/** Static component of the salt used to compute short txids for transaction reconciliation. */
```
A bit of fanciness, but you can insert a new line between each variable. Ease the reading IMHO.

naumenkogs replied:
We never do this is other places, so I'm leaning towards keeping it with no new lines, unless someone else also asks for it.

ariard commented on src/txreconciliation.cpp:
```diff
@@ -0,0 +1,893 @@
+
+    /** Keep track of reconciliations with the peer. */
+    ReconciliationInitByUs m_state_init_by_us;
+    ReconciliationInitByThem m_state_init_by_them;
```
What purpose serve those 2 different structures ? Can't they be aggregated in one ?

naumenkogs replied:
They have very different internals, and I think it's for ease of understanding. If we are initiator, we don't care about those fields in other struct.

Both of them could have been inheritance child classes of ReconciliationState (but with extra fields) and then we have only one in m_states, but I think in C++ specifically it's easier to implement them as extension classes like this.

ariard replied:
Okay I took the point around ease of understanding. Maybe you can document around `m_state_init_by_us`/`m_state_init_by_them` in `ReconciliationState` why in protocol flow only one of them matters.

ariard commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,134 @@
+     * If the peer was not previously registered for reconciliations or it's not the time yet,
+     * returns nullopt.
+     */
+    std::optional<std::vector<uint8_t>> MaybeRespondToReconciliationRequest(const NodeId peer_id);
```
IMO, it might be better to pass the vector by reference as an argument instead of returning a `std::optional`. I believe same comment can be made for `MaybeRequestReconciliation`/`HandleSketch`/`FinalizeIncomingReconciliation`.

naumenkogs replied:
And then return bool? Or return void and make vector empty in case of failure? Maybe depending on the usecase.

ariard replied:
Yes returning a bool. IMHO, boolean return make it easier to read code paths.

ariard commented on src/txreconciliation.h:
```diff
@@ -0,0 +1,134 @@
+
+    /**
+     * Returns whether a given peer might respond to our reconciliation requests.
+     * If the peer was not previously registered for reconciliations, returns nullopt.
```
I think you can drop `std::optional` for `IsPeerResponder`/`GetPeerSetSize`/... you can tread absence of reconciliation registration as a failure (=false) directly.

You might even drop `IsPeerRegistered` completely, currently it always precedes calls to other `TxReconciliationTracker` methods so not really an independent method.

naumenkogs replied:
>I think you can drop std::optional for IsPeerResponder/GetPeerSetSize/... you can tread absence of reconciliation registration as a failure (=false) directly.

Currently, IsPeerResponder=true if we initiate and false if they initiate.

If peer is not even registered, it's *neither*. It's not just false (maybe currently default=false can be fine because we have only 1 call, but I think it's confusing, no?)

naumenkogs replied:
>You might even drop IsPeerRegistered completely, currently it always precedes calls to other TxReconciliationTracker methods so not really an independent method.

For the same reason explained above, I think this is currently useful.

ariard replied:
Okay after looking further on code paths where `IsPeerResponder` used it make sense in some cases to reduce ambiguity. E.g if `GetPeerSetSize` would return a `recon_set_size=0` you will try to reconciliate instead of backing off.

ariard commented on src/txreconciliation.cpp:
```diff
@@ -0,0 +1,893 @@
+        return sketch;
+    }
+
+    public:
```
Do you need to make those `ReconciliationState` methods public ?

naumenkogs replied:
Currently, nothing has to be made public, because it's a friend class. So you think everything should be private?

ariard replied:
Yes I think it should be private by default. In the future, if a use-case arise, we can make them public.

ariard reviewed (commented):
Thanks for the effort around encapsulation, it's far better :) Some last minor comments.

naumenkogs committed:
Add helper to see tx relay support by conn type

We clearly know which peers are completely incapable of
transaction relay based on the connection type, and which
might (at least in theory) relay transactions. Add a function
to distinguish between the two.

This helper will be later used to see whether we should
announce tx reconciliation support to the peer, and also
to see if we reached a limit of peers we flood to (per Erlay).

naumenkogs committed:
Announce reconciliation support

If we're connecting to the peer which might support
transaction reconciliation, we announce or desire to
do reconciliations with them, along with the
negotiation parameters.

We store the reconciliation salt so that when the peer
responds with their salt, we are able to compute the
full recon salt.

naumenkogs committed:
Add helper to compute reconciliation salt

Transaction reconciliation assumes exchanging
sketches representing transactions a peer wants
to announce. Transactions are included in sketches
based on the short ID, which is computed by hashing
with the use of salt.

naumenkogs committed:
Finish negotiating reconciliation support

Once we received a reconciliation announcement support
message from a peer and it doesn't violate our protocol,
we store the negotiated parameters which will be used
for future reconciliations.

naumenkogs committed:
Add helper to see if a peer reconciles txs with us

naumenkogs committed:
Keep track of outbound peers which we flood txs to

This will be later used to see if we reached the limit
(to be added) on outbound flood conns.

naumenkogs committed:
Assign some outbound peers we reconcile with for flooding

To facilitate fast transaction relay when reconciliation
is broadly deployed while still conserving bandwidth,
we will flood certain transactions to a subset of outbound
peers with reconciliation enabled.

Here we choose those peers at handshake time in case
if we don't have enough of those peers yet.

naumenkogs committed:
Remove peer's reconciliation state on disconnect

naumenkogs committed:
Add helper to see we are recon initiator for the peer

naumenkogs committed:
Assign announcement protocol based on tx source

If a transaction came from an inbound, flood.
This would facilitate rapid relay across the backbone
of reachable nodes with limited bandwidth.

If a transaction came from an outbound, reconcile.
This would allow to conserve most of the bandwidth.

If a transaction originated locally, reconcile.
This would avoid making obvious which transactions
originated at non-reachable nodes.

naumenkogs committed:
Add peers to reconciliation queue on negotiation

When we're finalizing negotiation, we should add the peers
for which we will initiate reconciliations to the queue.

naumenkogs committed:
Track tx announcement protocol

Some transactions will be flooded to peers we
reconcile with. Facilitate tracking those transactions
with a flag.

naumenkogs committed:
Schedule a response to a reconciliation request

Do not respond to a request right away as it would enable
a DoS attack and allow monitoring of transaction a node has.
Instead, respond to all reconciliation requests in batch
after a small delay.

naumenkogs committed:
Handle reconciliation request

Store the parameters the peer sent us inside the
reconciliation request.

naumenkogs committed:
Respond to a reconciliation request

When the time comes, we should send a sketch of our
local reconciliation set to the reconciliation initiator.

naumenkogs committed:
Handle reconciliation sketch and successful decoding

naumenkogs committed:
Add helper to parse vector sketch representation

naumenkogs committed:
Add helper to compute sketch extensions

Sketch extension may be requested if our initial
sketch was not sufficient for the peer to find the
difference due to low capacity.

naumenkogs committed:
Prepare for sketch extension request

To be ready to respond to a sketch extension request
from our peer, we should store a snapshot of our state
and capacity of the initial sketch, so that we compute
extension of the same size and over the exact same
transactions.

Transactions arriving during this reconciliation will
be instead stored in the regular set.

naumenkogs committed:
Handle reconciliation extension request

If peer failed to reconcile based on our initial response sketch,
they will ask us for a sketch extension. Store this request to respond later.

naumenkogs committed:
Respond to sketch extension request

Sending an extension may allow the peer to reconcile
transactions, because now the full sketch has twice
as much capacity.

naumenkogs committed:
Handle sketch extension

If a peer sent us an extension sketch, we should
reconstruct a full sketch from it with the snapshot
we stored initially, and attempt to decode the difference.
